container-registry-configuration/README.md

```diff
@@ -0,0 +1,189 @@
+
+```
+[plugins."io.containerd.grpc.v1.cri".registry]
+
+[plugins."io.containerd.grpc.v1.cri".registry.mirrors]
+[plugins."io.containerd.grpc.v1.cri".registry.mirrors."docker.io"]
+endpoint = ["giantswarm.azurecr.io"]
+
+[plugins."io.containerd.grpc.v1.cri".registry.configs]
+[plugins."io.containerd.grpc.v1.cri".registry.configs."registry-1.docker.io".auth]
+username = "giantswarm-user-account"
+password = "my_token_from_docker_hub"
+```
+
+#### Decision
+
+`1.c` is selected.
+We will use containerd configuration. 
+
+### 2. Authentication
+
+#### 2.a Using unauthenticated accounts
+
+Using only public images without authentication is an option to avoid complexity of securing credentials. However, registry providers have pull/rate limits per account. We will be limited by total of [free account limit](https://docs.docker.com/docker-hub/download-rate-limit/) of Docker Hub (100 pulls per 6 hours per IP address) and [tier limit](https://learn.microsoft.com/en-us/azure/container-registry/container-registry-skus) of Azure Container Registry (3000 ReadOps per minute for standard). Anyone can pull images from our registry and spend our limits, which makes us vulnerable. 
+
+#### 2.b Using authenticated accounts
+
+We can pass credentials to `containerd` configuration to be not limited by registry provider limits.
+
+#### Decision
+
+`2.b` is selected.
+We will use authenticated accounts.
+
+
+### 3. Registries
+
+The intranet page (See `Registry Mirrors`) states this as a desired configuration:
+  - `docker.io` as primary public registry since it has a privilege in docker daemon.
+  - Our own Azure Container Registry as secondary registry because of the security concerns
+
+Since we use `containerd`, docker has not a privilege anymore but there is no need to change this order at the moment. We will follow this configuration for CAPI clusters too.
+
+### 4. Accounts
+
+- 4.a Giantswarm Account + Single account
+- 4.b Giantswarm Account + Per MC including its WCs
+- 4.c Giantswarm Account + Per WC
+- 4.b Customer Accounts
+
+#### Decision
+
+`4.b` is selected.
+
+Since it is a part of the platform itself, we are going to use Giant Swarm accounts in containerd configuration.
+As we do for other operators, we will follow "per MC" approach here to.
+
+### 5. How to provide credentials to WCs
+
+#### 5.a cluster-apps-operator
+
+We can propogate credentials from MC to WC by using `cluster-apps-operator`. It can create a secret per WC with the credentials.
+
+#### 5.b Optional interface + Gitops
+
+We can delegate the responsibility of creating credentials secret to the creator of `cluster-$provider` apps. GitOps can be useful to template the secrets per WC.
+
+#### Decision
+
+`5.b` is selected.
+
+We don't want to add another responsility to `cluster-apps-operator`. 
+
+
+### 6. Default configuration for WCs
+
+- 6.a: Configuring WCs by default
+- 6.b: Not configuring WCs by default
+
+#### Decision
+
+**NOT DECIDED YET**
```

Comment:
if we do not configure them by default we create unstable/unreliable clusters. In vintage we configure everything be default AFAIK. What would speak against that?

Reply:
To be able to do it by default, we need to create the credential secret in every organization.

Reply:
I think the clusters need to have the registry secret configured by default for the same reason Puja mentioned. The propagation of secrets across multiple namespaces could be handled by something like the ClusterSecret CR. There's an operator for this: https://github.com/zakkg3/ClusterSecret (but, I don't know how reliable it is).

Reply:
As discussed in last KaaS sync and with my background as having been part of the team designing the app interface, I'd really suggest to use the systems that are in place in the interface that honeybadger is providing us with the app platform. An MC-wide default is supposed to go into the catalog config/secret, it can be overridden by cluster with the cluster secret (and/or user secret), but that should be an optional override if needed.

I know there was visibility concerns on this but that is feedback for honeybadger that you can bring there to discuss.

Building operators and workaround to get functionality that is already there sounds like busy work to me. We can improve on what we have, but let's use it first to see if it creates issues and then we can see how to solve those issues. Visibility is a UX issue that does not put the solution in question IMHO.

Reply:
I didn't mean that we need to build a new operator (there is already one). I would just like to have the behavior of the ClusterSecret CR I mentioned. It's great if the app platform is able to do it and that we don't have to do anything else (no new code, no additional deployments).

In the end, I would like to see the following:
- default registry config (secret + url) per MC;
- this default is made available to all WCs (probably the responsibility of the app platform);
- `cluster-$provider` chart has default reference to the registry config (secret + url);
- `cluster-$provider` has the option to have a different registry config passed in (secret + url);
- `cluster-$provider` renders the `containerd` config;

The registry config can be a list of registries.

The `cluster-$provider` should also have the ability to have additional registries configured by the customer (appended to our defaults). Probably not needed now, but the design should not prevent it.
I can also see that in some private environments customers would want to have full control over the list of registries, but this requires more discussions.

Why? The defaults are needed in order to offer a good user experience out of the box. The configurability is needed to enable us to respond to customer requests and to allow us the flexibility needed in testing.